File: DataStructures/src/main/java/com/zhangdd/linkedlist/DoubleLinkedList.java
package com.zhangdd.linkedlist;

/**
 * @author zhangdd on 2021/9/29
 */
public class DoubleLinkedList {

    private Node head = new Node(0, null, null);

    public Node getHead() {
        return head;
    }

    public void list() {
        Node temp = head.next;
        while (temp != null) {
            System.out.println(temp);
            temp = temp.next;
        }
    }

    /**
     * 添加节点到链表最后
     */
    public void add(Node node) {
        Node temp = head;
        while (temp != null) {
            if (temp.next == null) {
                break;
            }
            temp = temp.next;
        }
        temp.next = node;
        node.pre = temp;
    }

    public void addByOrder(Node node) {
        Node temp = head;
        boolean findNodeFlag = false;
        while (temp != null) {
            if (temp.next == null) {
                findNodeFlag = true;
                break;
            }
            if (temp.next.no > node.no) {
                findNodeFlag = true;
                break;
            }
            if (temp.next.no == node.no) {
                break;
            }
            temp = temp.next;
        }
        if (findNodeFlag) {
            node.next = temp.next;
            node.pre = temp;
            temp.next = node;
        }


    }

    /**
     * 修改节点
     */
    public void update(Node node) {
        Node temp = head.next;

        boolean findNodeFlag = false;
        while (temp != null) {
            if (temp.no == node.no) {
                findNodeFlag = true;
                break;
            }
            temp = temp.next;
        }
        if (findNodeFlag) {
            temp.name = node.name;
            temp.nickname = node.nickname;
        }
    }

    /**
     * 删除一个节点
     */
    public void del(int no) {
        Node temp = head.next;
        boolean findNodeFlag = false;
        while (temp != null) {
            if (temp.no == no) {
                findNodeFlag = true;
                break;
            }
            temp = temp.next;
        }
        if (findNodeFlag) {
            temp.pre.next = temp.next;
            if (temp.next != null) {
                temp.next.pre = temp.pre;
            }
        }
    }


    public static class Node {
        private Node next;
        private Node pre;
        private int no;
        private String name;
        private String nickname;

        public Node(int no, String name, String nickname) {
            this.no = no;
            this.name = name;
            this.nickname = nickname;
        }

        @Override
        public String toString() {
            return "Node{" +
                    "no=" + no +
                    ", name='" + name + '\'' +
                    ", nickname='" + nickname + '\'' +
                    '}';
        }
    }
}

File: python/DataStructures/linkedlist/DoubleLinkedList.py
#!/usr/bin/env python3
"""
@author: zhangdd on 2022/1/13
双向链表
"""


class Node:
    def __init__(self, data, name):
        self.data = data
        self.name = name
        self.pre = None
        self.next = None

    def show(self):
        print('data:{},name:{}'.format(self.data, self.name))


class DoubleLinkedList:

    def __init__(self):
        self.head = Node(-1, None)

    def add_node(self, node):
        tmp = self.head
        while tmp.next is not None:
            tmp = tmp.next
        tmp.next = node
        node.pre = tmp

    def del_node(self, data):
        tmp = self.head.next
        while tmp is not None:
            if tmp.data == data:
                tmp.pre.next = tmp.next
                tmp.next.pre = tmp.pre
                break
            tmp = tmp.next

    def show(self):
        tmp = self.head
        while tmp.next is not None:
            tmp.next.show()
            tmp = tmp.next


link_list = DoubleLinkedList()
node1 = Node(1, '鲁迅')
node2 = Node(2, '鲁肃')
node3 = Node(3, '周瑜')
node4 = Node(4, '黄盖')

link_list.add_node(node1)
link_list.add_node(node2)
link_list.add_node(node3)
link_list.add_node(node4)

link_list.show()

print('-' * 50)

link_list.del_node(3)
link_list.del_node(2)

link_list.show()
